Pseudo PR for nf-core/scrnaseq: 1.0.0 Release

apeltzer opened the pull request:
Hi everyone,

this marks functional testing for all three pipeline options:

- STARSolo
- Kallisto+Bustools
- Salmon Alevin + AlevinQC 

Please do a full review and we can heopefully get this done / sorted asap. Will also require a bit more docs updates, but the general code should be almost ready to go with lots of contributions from @PeterBailey , @olgabot and me. 


## PR checklist
 - [x] This comment contains a description of changes (with reason)
 - [x] If you've fixed a bug or added code that should be tested, add tests!
 - [x] If necessary, also make a PR on the [nf-core/scrnaseq branch on the nf-core/test-datasets repo]( https://github.com/nf-core/test-datasets/pull/new/nf-core/scrnaseq)
 - [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
 - [x] Make sure your code lints (`nf-core lint .`).
 - [x] Documentation in `docs` is updated
 - [x] `CHANGELOG.md` is updated
 - [x] `README.md` is updated

This is a pseudo PR - please just use this to do a full review of everything. Note, that the intiial template used wa snf-core v1.5, so there are quite a lot of changes since then 😓

MaxUlysse commented on .github/CONTRIBUTING.md:
```diff
@@ -6,9 +6,10 @@ We try to manage the required tasks for nf-core/scrnaseq using GitHub issues, yo
 However, don't be put off by this template - other more general issues and suggestions are welcome! Contributions to the code are even more welcome ;)
 
-> If you need help using or modifying nf-core/scrnaseq then the best place to go is the Gitter chatroom where you can ask us questions directly: https://gitter.im/nf-core/Lobby
+> If you need help using or modifying nf-core/scrnaseq then the best place to ask is on the pipeline channel on [Slack](https://nf-core-invite.herokuapp.com/).
```
```suggestion
> If you need help using or modifying nf-core/scrnaseq then the best place to ask is on the pipeline channel on [Slack](https://nfcore.slack.com/channels/scrnaseq).
```

MaxUlysse commented on .github/CONTRIBUTING.md:
```diff
@@ -22,24 +23,27 @@ is as follows:
 ## Getting help
-For further information/help, please consult the [nf-core/scrnaseq documentation](https://github.com/nf-core/scrnaseq#documentation) and don't hesitate to get in touch on [Gitter](https://gitter.im/nf-core/Lobby)
+
+For further information/help, please consult the [nf-core/scrnaseq documentation](https://github.com/nf-core/scrnaseq#documentation) and don't hesitate to get in touch on the pipeline channel on [Slack](https://nf-core-invite.herokuapp.com/).
```
```suggestion
For further information/help, please consult the [nf-core/scrnaseq documentation](https://github.com/nf-core/scrnaseq#documentation) and don't hesitate to get in touch on the pipeline channel on [Slack](https://nfcore.slack.com/channels/scrnaseq).
```

MaxUlysse commented on CODE_OF_CONDUCT.md:
```diff
@@ -34,7 +34,7 @@ This Code of Conduct applies both within project spaces and in public spaces whe
 ## Enforcement
 
-Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting the project team on the [Gitter channel](https://gitter.im/nf-core/Lobby). The project team will review and investigate all complaints, and will respond in a way that it deems appropriate to the circumstances. The project team is obligated to maintain confidentiality with regard to the reporter of an ...
+Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting the project team on [Slack](https://nf-core-invite.herokuapp.com/). The project team will review and investigate all complaints, and will respond in a way that it deems appropriate to the circumstances. The project team is obligated to maintain confidentiality with regard to the reporter of an incide ...
```
```suggestion
Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting the project team on [Slack](https://nf-co.re/join/slack). The project team will review and investigate all complaints, and will respond in a way that it deems appropriate to the circumstances. The project team is obligated to maintain confidentiality with regard to the reporter of an incident. Further details of specific enforcement policies may be posted separately.
```

MaxUlysse commented on README.md:
```diff
@@ -1,30 +1,42 @@
 
 [![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg)](http://bioconda.github.io/)
 [![Docker](https://img.shields.io/docker/automated/nfcore/scrnaseq.svg)](https://hub.docker.com/r/nfcore/scrnaseq)
+[![Install with Singularity](https://img.shields.io/badge/use%20with-singularity-purple.svg)](https://www.sylabs.io/docs/)
```
Do we do that still?

apeltzer replied:
No - will drop this here.

MaxUlysse commented on README.md:
```diff
@@ -1,30 +1,42 @@
 [![Docker](https://img.shields.io/docker/automated/nfcore/scrnaseq.svg)](https://hub.docker.com/r/nfcore/scrnaseq)
+[![Install with Singularity](https://img.shields.io/badge/use%20with-singularity-purple.svg)](https://www.sylabs.io/docs/)
+
+[![Join us on Slack](https://img.shields.io/badge/slack-nfcore/scrnaseq-blue.svg)](https://nfcore.slack.com/messages/CHN5BV5DW)
```
```suggestion
[![Join us on Slack](https://img.shields.io/badge/slack-nfcore/scrnaseq-blue.svg)](https://nfcore.slack.com/channels/scrnaseq)
```

apeltzer commented on README.md:
```diff
@@ -1,30 +1,42 @@
 
 [![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg)](http://bioconda.github.io/)
 [![Docker](https://img.shields.io/docker/automated/nfcore/scrnaseq.svg)](https://hub.docker.com/r/nfcore/scrnaseq)
+[![Install with Singularity](https://img.shields.io/badge/use%20with-singularity-purple.svg)](https://www.sylabs.io/docs/)
```
```suggestion

```

MaxUlysse commented on conf/test_star.config:
```diff
@@ -0,0 +1,30 @@
+ *   nextflow run nf-core/scrnaseq -profile test
+ */
+
+params {
```
This file looks very similar to `test.config`.
Did you try simplifying this one and including it?

apeltzer replied:
Now I did ;-)

MaxUlysse commented on docs/usage.md:
```diff
@@ -13,9 +13,18 @@
+  * [`--salmon_index`](#--salmon_index)
+  * [`--txp2gene`](#--txp2gene)
+  * [`--kallisto_gene_map`](#--kallisto_gene_map)
   * [`--igenomesIgnore`](#--igenomesignore)
```
Don't we have `igenomes_ignore` now?

MaxUlysse commented on docs/usage.md:
````diff
@@ -49,13 +58,12 @@ It is recommended to limit the Nextflow Java virtual machines memory. We recomme
 
 ```bash
-nextflow run nf-core/scrnaseq --reads '*_R{1,2}.fastq.gz' -profile docker
+nextflow run nf-core/scrnaseq --reads '*_R{1,2}.fastq.gz' --type '10x' --aligner 'kallisto' --barcode_whitelist 10xv2.whitelist.txt --fasta human.fasta --gtf human.gtf -profile docker
````
Do we really need quotes around `kallisto`?
Typical command is run with Kallisto then I suppose?
Should it not be specified by default in the config files? (with the specific files)

MaxUlysse commented on docs/usage.md:
```diff
@@ -65,28 +73,30 @@ Note that the pipeline will create the following files in your working directory
+
 It's a good idea to specify a pipeline version when running the pipeline on your data. This ensures that a specific version of the pipeline code and software are used when you run your pipeline. If you keep using the same tag, you'll be running the same version of the pipeline, even if there have been changes to the code since.
 
 First, go to the [nf-core/scrnaseq releases page](https://github.com/nf-core/scrnaseq/releases) and find the latest version number - numeric only (eg. `1.3.1`). Then specify this when running the pipeline with `-r` (one hyphen) - eg. `-r 1.3.1`.
```
```suggestion
First, go to the [nf-core/scrnaseq releases page](https://github.com/nf-core/scrnaseq/releases) and find the latest version number - numeric only (eg. `1.0.0`). Then specify this when running the pipeline with `-r` (one hyphen) - eg. `-r 1.0.0`.
```

MaxUlysse replied:
Should be good to mention a release that we actually have, we should have a TODO for that in the template I think.

apeltzer replied:
First one will be 1.0.0 - I adjusted this accordingly now :-)

MaxUlysse commented on docs/usage.md:
```diff
@@ -132,12 +142,46 @@ By default, the pipeline expects paired-end data. If you have single-end data, y
+* Salmon Alevin + AlevinQC
+* STARsolo
+
+To choose which one to use, please specify either `alevin`, `star` or `kallisto` as a parameter option for `--aligner`.
```
Is there a default one?

MaxUlysse commented on docs/usage.md:
````diff
@@ -170,69 +212,106 @@ params {
 --fasta '[path to Fasta reference]'
 ```
 
+> Note that you need to specify either a `--genome` or `--fasta` when running the STARsolo workflow. The Kallisto and Alevin workflows can utilize a `--transcript_fasta` instead, whereas STAR needs a genomic FastA file as input in all cases.
````
```suggestion
> Note that you need to specify either a `--genome` or `--fasta` when running the STARsolo workflow. The Kallisto and Alevin workflows can utilize a `--transcript_fasta` instead, whereas STAR needs a genomic fasta file as input in all cases.
```

MaxUlysse replied:
isn't it fasta files? or is it actually FastA files? Have I been lied to my whole life?

MaxUlysse commented on docs/usage.md:
```diff
@@ -170,69 +212,106 @@ params {
+
+### `--transcript_fasta`
+
+If you intend to skip the generation of a transcriptomic FastA file, you can use this parameter to supply a transcriptomic FastA file here. If you don't specify this, it will be automatically generated from the supplied genomics FastA file utilizing the GTF annotation subsequently.
```
same question here.
`fasta` or `FastA`?

MaxUlysse commented on docs/usage.md:
```diff
@@ -170,69 +212,106 @@ params {
 If you are likely to be running `nf-core` pipelines regularly it may be a good idea to request that your custom config file is uploaded to the `nf-core/configs` git repository. Before you do this please can you test that the config file works with your pipeline of choice using the `-c` parameter (see definition below). You can then create a pull request to the `nf-core/configs` repository with th ...
 
-If you have any questions or issues please send us a message on [`Slack`](https://nf-core-invite.herokuapp.com/).
+If you have any questions or issues please send us a message on [Slack](https://nf-core-invite.herokuapp.com/).
```
```suggestion
If you have any questions or issues please send us a message on [Slack](https://nfcore.slack.com/channels/scrnaseq/).
```

MaxUlysse commented on nextflow.config:
```diff
@@ -9,22 +9,47 @@
+  transcriptome_fasta = false
+  gtf = false
   singleEnd = false
+  aligner = 'alevin'
```
ok, so default aligner is `alevin` then

MaxUlysse commented on main.nf:
```diff
@@ -176,57 +248,479 @@ ${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style
 }
 
+if(params.type == '10x' && !params.barcode_whitelist){
+process unzip_10x_barcodes {
```
Indentation is wrong

MaxUlysse commented on main.nf:
```diff
@@ -176,57 +248,479 @@ ${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style
+
+if (params.aligner == 'star' && !params.star_index && params.fasta){
+  process makeSTARindex {
+       label 'high_memory'
```
2 spaces and then 4 spaces indentation?

MaxUlysse commented on main.nf:
```diff
@@ -176,57 +248,479 @@ ${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style
+    label 'mid_memory'
+    publishDir "${params.outdir}/kallisto/raw_bus", mode: 'copy'
+
+    when:
```
`when` statement should be after `output`.
Then you could remove the if

MaxUlysse commented on main.nf:
```diff
@@ -176,57 +248,479 @@ ${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style
+    publishDir "${params.outdir}/kallisto/sort_bus", mode: 'copy'
+
+    when:
+    params.aligner == "kallisto" && !params.skip_bustools
```
You should just need `!params.skip_bustools` in your `when` statement, since channel `kallisto_bus_to_sort` is chained from process `kallisto` it depends from the previous process, but it should be after output.

MaxUlysse commented on main.nf:
```diff
@@ -176,57 +248,479 @@ ${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style
+    publishDir "${params.outdir}/kallisto/bustools_counts", mode: "copy"
+
+    when:
+    params.aligner == 'kallisto' && !params.skip_bustools
```
You should not need any `when` statement here

MaxUlysse commented on main.nf:
```diff
@@ -176,57 +248,479 @@ ${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style
+    tag "$bus"
+    publishDir "${params.outdir}/kallisto/bustools_metrics", mode: "copy"
+
+    when:
```
You should not need any `when` statement here

MaxUlysse commented on main.nf:
```diff
@@ -176,57 +248,479 @@ ${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style
+    label 'high_memory'
+    publishDir "${params.outdir}/alevin", mode: 'copy'
+
+    when:
```
Don't need that one

MaxUlysse commented:
From what I understood `alevin` is the default aligner, but it should be mention in the docs, and I think we can improve the tests.
But that's not vital.
All `when` statement are not position after `output`, so don't think any of them actually works.
Which I guess is why there is so many if logic involved.
If you can fix all that, I'm sure it will simplify workflow logic itself.

apeltzer commented:
Thanks for all the feedback @MaxUlysse !

* Improved and replaced most of the if statements with proper when where required
* dropped most of the when statements whenever (hehe) there is a chain of processes not requiring it
* Lots of docs for outputs, usage and handling of the pipeline (e.g. defaults explained)
* Cleaned up tests, only two profiles now as they use different datasets for testing :-)
